portability/toku_pthread.h

```diff
@@ -217,40 +218,125 @@ toku_mutex_assert_unlocked(toku_mutex_t *mutex) {
     invariant(!mutex->locked);
 }
 #else
-static inline void
-toku_mutex_assert_unlocked(toku_mutex_t *mutex __attribute__((unused))) {
+static inline void toku_mutex_assert_unlocked(toku_mutex_t *mutex
+                                              __attribute__((unused))) {}
+#endif
+
+#define toku_mutex_lock(M) \
+    toku_mutex_lock_with_source_location(M, __FILE__, __LINE__)
+
+static inline void toku_cond_init(toku_cond_t *cond,
+                                  const toku_pthread_condattr_t *attr) {
+    int r = pthread_cond_init(&cond->pcond, attr);
+    assert_zero(r);
 }
+
+#define toku_mutex_trylock(M) \
+    toku_mutex_trylock_with_source_location(M, __FILE__, __LINE__)
+
+inline void toku_mutex_unlock(toku_mutex_t *mutex) {
+#if TOKU_PTHREAD_DEBUG
+    invariant(mutex->owner == pthread_self());
+    invariant(mutex->valid);
+    invariant(mutex->locked);
+    mutex->locked = false;
+    mutex->owner = 0;
 #endif
+    toku_instr_mutex_unlock(mutex->psi_mutex);
+    int r = pthread_mutex_unlock(&mutex->pmutex);
+    assert_zero(r);
+}
 
-typedef struct toku_cond {
-    pthread_cond_t pcond;
-} toku_cond_t;
+inline void toku_mutex_lock_with_source_location(toku_mutex_t *mutex,
+                                                 const char *src_file,
+                                                 int src_line) {
+    int r = 0;
 
-// Same considerations as for ZERO_MUTEX_INITIALIZER apply
-#define ZERO_COND_INITIALIZER {}
+    toku_mutex_instrumentation mutex_instr;
+    toku_instr_mutex_lock_start(mutex_instr, *mutex, src_file, src_line);
 
-#define TOKU_COND_INITIALIZER {PTHREAD_COND_INITIALIZER}
+    r = pthread_mutex_lock(&mutex->pmutex);
+    toku_instr_mutex_lock_end(mutex_instr, r);
 
-static inline void
-toku_cond_init(toku_cond_t *cond, const toku_pthread_condattr_t *attr) {
+    assert_zero(r);
+#if TOKU_PTHREAD_DEBUG
+    invariant(mutex->valid);
+    invariant(!mutex->locked);
+    invariant(mutex->owner == 0);
+    mutex->locked = true;
+    mutex->owner = pthread_self();
+#endif
+}
+
+inline int toku_mutex_trylock_with_source_location(toku_mutex_t *mutex,
+                                                   const char *src_file,
+                                                   int src_line) {
+    int r = 0;
+    toku_mutex_instrumentation mutex_instr;
+    toku_instr_mutex_trylock_start(mutex_instr, *mutex, src_file, src_line);
+
+    r = pthread_mutex_lock(&mutex->pmutex);
```

Comment:
Declare at initialization:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-always
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-introduce
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-init

const:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-const

I don't know if new TokuDB code should follow C++ Core Guidelines, IMHO they should. Maybe this is something you discussed at the dev meeting or can discuss now with @georgelorchpercona now.